packages/compartment-mapper/src/policy.js

```diff
@@ -440,39 +440,85 @@ async function attenuateModule({
   // For async attenuators see PR https://github.com/endojs/endo/pull/1535
 
-  return freeze({
-    imports: originalModuleRecord.imports,
+  return {
+    imports: moduleSource.imports,
     // It seems ok to declare the exports but then let the attenuator trim the values.
     // Seems ok for attenuation to leave them undefined - accessing them is malicious behavior.
-    exports: originalModuleRecord.exports,
+    exports: moduleSource.exports,
     execute: (moduleExports, compartment, resolvedImports) => {
       const ns = {};
-      originalModuleRecord.execute(ns, compartment, resolvedImports);
+      moduleSource.execute(ns, compartment, resolvedImports);
       const attenuated = attenuate(ns);
       moduleExports.default = attenuated;
       assign(moduleExports, attenuated);
     },
-  });
+  };
+}
+
+/**
+ * Attenuates a module
+ * @param {object} options
+ * @param {import('./types.js').DeferredAttenuatorsProvider} options.attenuators
+ * @param {import('./types.js').AttenuationDefinition} options.attenuationDefinition
+ * @param {import('ses').VirtualModuleSource | import('ses').StrictModuleDescriptor} options.moduleDescriptor
+ * @returns {Promise<import('ses').StrictModuleDescriptor>}
+ */
+async function attenuateModule({
+  attenuators,
+  attenuationDefinition,
+  moduleDescriptor,
+}) {
+  await null;
+  if ('source' in moduleDescriptor) {
+    const { source: moduleSource } = moduleDescriptor;
+    if (
+      typeof moduleSource !== 'string' &&
+      'imports' in moduleSource &&
+      'exports' in moduleSource &&
+      'execute' in moduleSource
+    ) {
+      return {
+        source: await attenuateVirtualModuleSource({
+          attenuators,
+          attenuationDefinition,
+          moduleSource,
+        }),
+      };
+    }
+  } else if (
+    'imports' in moduleDescriptor &&
+    'exports' in moduleDescriptor &&
+    'execute' in moduleDescriptor
+  ) {
+    return {
+      source: await attenuateVirtualModuleSource({
+        attenuators,
+        attenuationDefinition,
+        moduleSource: moduleDescriptor,
+      }),
+    };
+  }
+  throw new Error('Can only attenuate virtual module source descriptors');
 }
 
 /**
  * Throws if importing of the specifier is not allowed by the policy
  *
  * @param {string} specifier - exit module name
- * @param {import('ses').ThirdPartyStaticModuleInterface} originalModuleRecord - reference to the exit module
+ * @param {import('ses').VirtualModuleSource | import('ses').StrictModuleDescriptor} moduleDescriptor - reference to the exit module
  * @param {import('./types.js').PackagePolicy} policy - local compartment policy
  * @param {import('./types.js').DeferredAttenuatorsProvider} attenuators - a key-value where attenuations can be found
- * @returns {Promise<import('ses').ThirdPartyStaticModuleInterface>} - the attenuated module
+ * @returns {Promise<import('ses').ModuleDescriptor>} - the attenuated module descriptor
  */
 export const attenuateModuleHook = async (
```

Comment:
To what extent is this an externally visible API change?

I don't see any corresponding changes in tests, which suggests it's _not_ an external API change.
But that could be just a lack of test coverage. Or maybe it's only tested indirectly.

Reply:
As part of convergence with XS, VirtualModuleSource is an alias for ThirdPartyStaticModuleInterface and ModuleDescriptor is a new type that captures all of the new and old types importHooks may return (and the range of types that an importHook adapter is expected to handle). The change here allows the full range of module descriptors to pass through this adapter when a LavaMoat policy is absent, which allows us to surface the whole module descriptor range in an exit module importHook. This was a bottleneck before.

So, the input range expands from VirtualModuleSource to also accept StrictModuleDescriptor, the forward-compatible subset of ModuleDescriptor. I did not feel the need to create a new obligation for backward compatibility past the next SES major version and wasn’t able to get TypeScript to play along with reliably differentiating a ModuleExportsNamespace from a VirtualModuleSource. In SES, we use a brand check for that.

This catches us up with what I expected to be working behavior, so is more like a bugfix than a feature, so I didn’t capture this in the notes.

Reply:
Was this API in any previous release?